vpr/src/route/parallel_connection_router.cpp

```diff
@@ -341,52 +425,39 @@ void ParallelConnectionRouter::timing_driven_route_connection_from_heap_thread_f
                                                                          const t_bb& bounding_box,
                                                                          const size_t thread_idx) {
     // cheapest t_heap in current route tree to be expanded on
-    pq_node_t cheapest;
+    float new_total_cost;
+    RRNodeId inode;
     // While the heap is not empty do
-    while (heap_.try_pop(cheapest)) {
+    while (heap_.try_pop(new_total_cost, inode)) {
         // update_router_stats(router_stats_,
         //                     false,
         //                     cheapest->index,
         //                     rr_graph_);
 
-        RRNodeId inode = cheapest.index;
-
-        // Exit Condition
-        // if (inode == sink_node) {
-        //     update_cheapest(&cheapest, &rr_node_route_inf_[inode]);
-        //     break;
-        // }
-
-        float new_total_cost = cheapest.cost;
-        float new_back_cost = cheapest.backward_path_cost;
-        RREdgeId new_prev_edge = cheapest.prev_edge();
         // Pruning
-        // NOTE: This is thread-safe since, even though the values may change during operation, that is ok,
-        //       since if any of the arguments need to be pruned it will.
-        if (prune_node(inode, new_total_cost, new_back_cost, new_prev_edge, sink_node, rr_node_route_inf_, cost_params)) {
+        if (inode == sink_node) {
             continue;
         }
 
-        // Synchronization Point
-        obtainSpinLock(inode);
-
-        // Need to double check pruning since things may have changed since obtaining the lock.
-        if (prune_node(inode, new_total_cost, new_back_cost, new_prev_edge, sink_node, rr_node_route_inf_, cost_params)) {
-            releaseLock(inode);
+        if (should_not_explore_neighbors(inode, new_total_cost, sink_node, rr_node_route_inf_, cost_params)) {
             continue;
         }
 
-        // Update the global values now that we are in the critical section.
-        update_cheapest(&cheapest, &rr_node_route_inf_[inode], thread_idx);
+        obtainSpinLock(inode);
+
+        node_t cheapest;
+        cheapest.backward_path_cost = rr_node_route_inf_[inode].backward_path_cost;
+        cheapest.R_upstream = rr_node_route_inf_[inode].R_upstream;
+        cheapest.prev_edge = rr_node_route_inf_[inode].prev_edge;
 
         releaseLock(inode);
 
-        if (inode == sink_node) {
+        if (should_not_explore_neighbors(inode, new_total_cost, cheapest.backward_path_cost, sink_node, rr_node_route_inf_, cost_params)) {
```

Comment:
I think you should use the EXACT same call to should_not_explore_neighbors. This second one is not catching anything and leads to a lot of duplicate code. The basic idea is as follows:
1. Check that this node is worth exploring
2. Obtain the lock (may have to wait), update the global values, release the lock
3. Double check that this node is STILL worth exploring (using the exact same call as 1 above).

The third point above *may* not be necessary, so I would add a TODO above saying that we should try testing without it and see if anything changes.

Reply:
But, for the third point, the global values (cost) could be updated between 1 and 2 (during the waiting time for obtaining the lock), right?